algo.lua

```diff
@@ -0,0 +1,137 @@
+local ffi = require 'ffi'
+local errcheck = cudnn.errcheck
+
+local algo = {}
+local autotunerCache = {}
+autotunerCache[1] = {} -- forward
+autotunerCache[2] = {} -- backwardFilter
+autotunerCache[3] = {} -- backwardData
+
+local function setupAlgo(self, algo_t, perf_t, findAPI, getAPI, wsAPI, algSearchMode, params)
+
+        local algType = ffi.new(algo_t, 1)
+
+        if cudnn.benchmark or cudnn.fastest then -- the manual auto-tuner is run
+            if autotunerCache[1][self.autotunerHash] then
```

Comment:
So if you've already called setupAlgo for forward for a particular convolution, and are now calling it for dgrad or wgrad,  autotunerCache[**1**][self.autotunerHash] will be non-nil, and you will return best forward algo for dgrad and wgrad? That won't work. If I'm not missing anything here, can you please fix it?